File: code/go/ch20/src/luago/number/math.go
package number

import "math"

// todo: correct?
func FloatToInteger(f float64) (int64, bool) {
	i := int64(f)
	return i, float64(i) == f
}

// a % b == a - ((a // b) * b)
func IMod(a, b int64) int64 {
	return a - IFloorDiv(a, b)*b
}

// a % b == a - ((a // b) * b)
func FMod(a, b float64) float64 {
	if a > 0 && math.IsInf(b, 1) || a < 0 && math.IsInf(b, -1) {
		return a
	}
	if a > 0 && math.IsInf(b, -1) || a < 0 && math.IsInf(b, 1) {
		return b
	}
	return a - math.Floor(a/b)*b
}

func IFloorDiv(a, b int64) int64 {
	if a > 0 && b > 0 || a < 0 && b < 0 || a%b == 0 {
		return a / b
	} else {
		return a/b - 1
	}
}

func FFloorDiv(a, b float64) float64 {
	return math.Floor(a / b)
}

func ShiftLeft(a, n int64) int64 {
	if n >= 0 {
		return a << uint64(n)
	} else {
		return ShiftRight(a, -n)
	}
}

func ShiftRight(a, n int64) int64 {
	if n >= 0 {
		return int64(uint64(a) >> uint64(n))
	} else {
		return ShiftLeft(a, -n)
	}
}

File: code/csharp/ch08/luavm/number/Math.cs
using System;

namespace luavm.number
{
    public class Math
    {
        internal static long IFloorDiv(long a, long b)
        {
            if (a > 0 && b > 0 || a < 0 && b < 0 || a % b == 0)
            {
                return a / b;
            }

            return a / b - 1;
        }

        internal static double FFloorDiv(double a, double b)
        {
            return System.Math.Floor(a / b);
        }

        internal static long IMod(long a, long b)
        {
            return a - IFloorDiv(a, b) * b;
        }

        internal static double FMod(double a, double b)
        {
            return a - System.Math.Floor(a / b) * b;
        }

        internal static long ShiftLeft(long a, long n)
        {
            if (n >= 0)
            {
                return a << (int) n;
            }

            return ShiftRight(a, (int) -n);
        }

        internal static long ShiftRight(long a, long n)
        {
            if (n >= 0)
            {
                return a >> (int) n;
            }

            return ShiftLeft(a, -n);
        }

        internal static Tuple<long, bool> FloatToInteger(double f)
        {
            var i = (long) f;
            return Tuple.Create(i, System.Math.Abs(i - f) <= 0);
        }
    }
}